### ISSUE TITLE
exp(p3): re-run PR #216 obs-fix + PR #225 MAPPO PPO training on post-#236 game with operative signaling

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#239]
## Context

**This is the scientifically critical re-derivation.** PR #236 turned the signal channel into an actual information-bearing input. PRs #216 (per-agent obs) and #225 (MAPPO centralized critic) were both trained on a substrate where signals were deterministic from work/rest — i.e., redundant noise.

The pre-#236 verdicts:
- PR #233 (#220 obs-fix): closes only **5.9%** of the minimal_specialization gap — verdict tier 3
- PR #232 (#231 MAPPO): all 3 scenarios tier 3, MAPPO marginally underperforms IPPO

**These verdicts may flip** when PPO can leverage a real signal channel. The deception research angle is now operative — agents in a Liar-containing team may rationally learn to distrust certain signals or weight them differently. That's exactly the kind of structure PPO was previously starved of.

## What to do

On `COMPUTE_HOST_PRIMARY`, repeat the protocols from issues #220 and #231:

1. **Obs-fix re-validation (sibling to #220)**:
   - Paired comparison: pre-#216 baseline vs post-#236 treatment (NOT post-#216, since post-#236 IS post-#216 + signal fix)
   - 2 scenarios × 3 seeds × 100 iters = 6 cells (treatment) + 6 cells (baseline if `minimal_specialization` not already covered)
   - Use `experiments/p3_specialization/analyze_220.py` against the new specialist/random baselines from sibling issues

2. **MAPPO re-validation (sibling to #231)**:
   - Paired sweep: 3 scenarios × 2 algos (IPPO/MAPPO) × 3 seeds × 100 iters = 18 cells
   - Use `experiments/p3_specialization/analyze_231.py`

3. **Verdict thresholds**: re-apply the pre-registered 3-tier ladder (>=50% gap closed = flip memory, 25-50% = partial win, <25% = still insufficient)

4. **If verdicts flip**: update `~/.claude/projects/-Users-rwalters-GitHub-bucket-brigade/memory/project_ppo_failure_mode.md` accordingly. The whole research narrative may shift.

## Dependencies

- [ ] #237 — Re-derive random baselines on post-#236 game (all 13 scenarios). **Required for `gap_closed` denominators.** Treatment-arm `gap_closed` is invalid until this lands.
- [ ] #238 — Re-derive specialist baselines on post-#236 game (`minimal_specialization`, `positional_default`, `default`). **Required for `gap_closed` numerators.** The hard-coded `MINSPEC_RANDOM = -96.07` / `MINSPEC_SPECIALIST = -22.07` in `analyze_220.py:53-54` and the `BASELINES` dict in `analyze_231.py:61-65` must be updated by #237/#238 before this issue's analyzer output is meaningful.

This issue can launch in parallel with #237/#238 on the same host (compute is the bottleneck, not artifact ordering — analyzer re-runs are cheap), but the **final verdict** cannot be applied until both sibling issues commit their updated baseline constants.

---

### Critical compatibility note: checkpoint loading after #236

**The action space changed from MultiDiscrete([10, 2]) → MultiDiscrete([10, 2, 2]).** Per the #236 PR body and curator review:

- **Obs dim is invariant at 46** (42 base + 4 identity tail for N=4). `last_actions` in the obs vector was deliberately kept at width 2 — the broadcast signal is exposed via `obs.signals`, not via re-encoding the action. So pre-#236 checkpoints have a compatible *input* layer.
- **Action heads change from 2 → 3.** `bucket_brigade/training/networks.py:58-60` iterates `action_dims` to build `self.action_heads`. A post-#236 policy will have 3 `nn.Linear` heads where pre-#236 checkpoints saved only 2.
- **This issue does NOT re-load pre-#236 checkpoints.** It trains fresh policies from scratch on post-#236 main against fresh random/specialist baselines (#237/#238). Pre-#236 checkpoints are useful only as a *historical* comparison and are out of scope.
- **The KL diagnostic needs an update.** `experiments/p3_specialization/diagnostics/pairwise_action_kl.py:37-43` hardcodes a 20-class joint (`p_house [B,10] × p_mode [B,2] → [B,20]`). Post-#236 the joint is 40-class (10×2×2). This must be fixed as part of this issue or the KL numbers will silently ignore the signal head. **Required change**: in `softmax_packed`, extend the outer product to include `p_signal = torch.softmax(logits_list[2], dim=-1)` and reshape to `[B, 40]`.

### Pre-flight smoke tests (before the real sweep)

Run these on the remote (with the post-#236 main checked out) to fail fast:

```bash
# 1. Confirm action space is 3D end-to-end
uv run python -c "
from bucket_brigade.envs import BucketBrigadeEnv
from bucket_brigade.envs.scenarios_generated import get_scenario_by_name
env = BucketBrigadeEnv(scenario=get_scenario_by_name('minimal_specialization', num_agents=4))
print('action_space:', env.action_space)  # expect MultiDiscrete([10, 2, 2])
print('obs flattened dim:', env.reset(seed=42))
"

# 2. Smoke 1-iter training cells to confirm action_dims propagate
uv run python -m experiments.p3_specialization.train \
  --scenario minimal_specialization --seed 42 \
  --num-iterations 1 --rollout-steps 256 \
  --output-dir /tmp/smoke_239_ippo
uv run python -m experiments.p3_specialization.train \
  --scenario minimal_specialization --seed 42 \
  --num-iterations 1 --rollout-steps 256 --centralized-critic \
  --output-dir /tmp/smoke_239_mappo

# 3. Verify saved checkpoint has 3 action heads
uv run python -c "
import torch
sd = torch.load('/tmp/smoke_239_ippo/policies/agent_0.pt', weights_only=True)
heads = [k for k in sd if 'action_heads' in k]
print('action_head keys:', heads)  # expect action_heads.0/1/2.{weight,bias}
"

# 4. Confirm KL diagnostic returns finite values for [10,2,2]
uv run python experiments/p3_specialization/diagnostics/pairwise_action_kl.py \
  --cell /tmp/smoke_239_ippo --no-write
```

### Run layout & tmux session naming

To avoid colliding with siblings #237 / #238 (both also targeting `COMPUTE_HOST_PRIMARY`), use these run-dir / tmux conventions:

| Sweep | Run dir | tmux session |
|---|---|---|
| Obs-fix baseline (pre-#216) | `experiments/p3_specialization/runs/issue239_obsfix_baseline/<scenario>/lambda_0e0/seed_<N>/` | `issue239_obsfix_baseline` |
| Obs-fix treatment (post-#236) | `experiments/p3_specialization/runs/issue239_obsfix_treatment/<scenario>/lambda_0e0/seed_<N>/` | `issue239_obsfix_treatment` |
| MAPPO sweep IPPO arm | `experiments/p3_specialization/runs/issue239_mappo/ippo/<scenario>/seed_<N>/` | `issue239_mappo_ippo` |
| MAPPO sweep MAPPO arm | `experiments/p3_specialization/runs/issue239_mappo/mappo/<scenario>/seed_<N>/` | `issue239_mappo_mappo` |

**Important**: the existing analyzers point at `issue220_{baseline,treatment}` and `issue231/{ippo,mappo}` paths. Either (a) write cells to those exact paths (clobbering pre-#236 runs — preferred since pre-#236 cells are scientifically stale), or (b) write to `issue239_*` and pass a path override / patch `_cell_dir` in both analyzers. Recommend (a) with a `git rev-parse HEAD` stamped into each cell's `config.json` for provenance.

### Parallelism strategy

- Mac Studio has ~10 perf cores; each cell is single-threaded CPU-bound.
- **MAPPO sweep (18 cells)**: 4-way parallel pool → ~4.5 cells per pool slot, ~14 min wall-clock at ~3 min/cell.
- **Obs-fix sweep (12 cells)**: 2-way parallel pool → 6 cells per slot, ~18 min wall-clock.
- **Total** (running both sweeps simultaneously with 6 concurrent cells): ~30 min.
- **Coordinate with siblings**: if #237 (random baseline re-derivation, ~30 min) and/or #238 (specialist, ~10 min) are launched first, this sweep should yield to them or share the host. Use `nohup` rather than tmux to be immune to tmux-server churn (see PR #233 notes).

### Expected wall-clock budget

- Pre-flight smoke tests: ~5 min (4 quick cells)
- MAPPO sweep (18 cells, 4-way parallel): ~14 min
- Obs-fix sweep (12 cells, 2-way parallel): ~18 min
- Analyzer re-runs (cheap, post-hoc on disk): <1 min
- Notebook amendment + memory update: 10 min
- **Total**: ~50 min single-agent serial, ~30 min if both sweeps parallel

### Gap_closed formula confirmation

From `experiments/p3_specialization/analyze_220.py:100-102` and `analyze_231.py:111-113`:

```python
def gap_closed(ppo_trailing5: float) -> float:
    return (ppo_trailing5 - random) / (specialist - random)
```

Where:
- `random` = mean per-step team reward over 50 episodes of `RandomAgent` × N (now emitting honest random signals after #236)
- `specialist` = mean per-step team reward of `SpecialistPolicy` × N (hand-tuned heuristic, honest signaling)
- `ppo_trailing5` = mean per-step team reward over the last 5 of `num_iterations` training iterations

**Both numerator and denominator change post-#236** because the substrate distribution shifts when one agent in a `RandomAgent`-mixed eval lies and another doesn't. Sibling issue #237 re-derives `random`; #238 re-derives `specialist`. **Do not run the analyzers until both land.**

### Where results land

- Per-cell artifacts: `experiments/p3_specialization/runs/issue239_*/...` (gitignored — kept on remote)
- Analyzer summaries (committed): `experiments/p3_specialization/diagnostics/results/{issue220_obsfix,issue231_mappo}/summary.{md,json}` — these get overwritten in place; existing pre-#236 summaries should be preserved as `summary.pre236.{md,json}` for the side-by-side comparison in the notebook.
- Dated notebook: new file `research_notebook/2026-05-16_issue239_post236_revalidation.md` with explicit pre-#236 vs post-#236 side-by-side tables.
- Memory file: `~/.claude/projects/-Users-rwalters-GitHub-bucket-brigade/memory/project_ppo_failure_mode.md` — update only if verdicts flip.

### Pre-registered verdict ladder

**Obs-fix (#220 pattern, single-scenario `minimal_specialization`)**:
- `gap_closed ≥ 0.50` → **PASS** (flip memory: "obs-fix sufficient with operative signal")
- `0.25 ≤ gap_closed < 0.50` → **PARTIAL** (signal helps but not decisive; queue follow-ups)
- `gap_closed < 0.25` → **STILL INSUFFICIENT** (memory line stays at "5.9% pre-#236, X% post-#236; bottleneck remains structural")

**MAPPO (#231 pattern, per-scenario `delta = gap_mappo - gap_ippo`)**:
- `delta ≥ 0.50` on ≥2 of 3 scenarios → **MAPPO_SUCCEEDS** (huge — flip memory)
- All 3 scenarios in `[0.25, 0.50)` → **MAPPO_HELPS_GLOBALLY_NOT_DECISIVE**
- Mixed tier-1 and tier-3 → **PARTIAL_SCENARIO_DEPENDENT**
- Any scenario `delta < -0.10` → **MAPPO_HARMFUL_ON_<SCEN>** (flagged, overrides)
- All 3 at tier-3 (delta < 0.25) → **INSUFFICIENT** (current pre-#236 verdict; memory unchanged)

### Risks

1. **Baseline drift mid-sweep**: if #237/#238 update the baselines after this issue's cells are on disk but before analyzer runs, the gap_closed math will use mismatched substrate. Mitigation: re-run analyzers AFTER both sibling issues have landed and their constants are merged to `main`.
2. **KL diagnostic stale**: as noted above, `pairwise_action_kl.py` hardcodes 20-class joints. If not fixed in this issue, the KL column in the analyzer summary will silently understate divergence. Required fix is small (~3 lines in `softmax_packed`).
3. **Memory invalidation under parallelism**: if siblings churn the same `research_notebook/` files, use distinct dated filenames per issue.
4. **Compute contention**: PR #233 noted tmux-session invalidation during parallel runs. Use `nohup` and detach cleanly.

## Affected Files

- `experiments/p3_specialization/train.py` — entry point for per-cell training; supports `--centralized-critic` flag (line 618).
- `experiments/p3_specialization/run_sweep.py` — grid driver (use for IPPO sweep).
- `experiments/p3_specialization/analyze_220.py` — obs-fix analyzer; uses `MINSPEC_RANDOM`/`MINSPEC_SPECIALIST` constants at lines 53-54 (will be updated by #237/#238).
- `experiments/p3_specialization/analyze_231.py` — MAPPO analyzer; uses `BASELINES` dict at lines 61-65 (will be updated by #237/#238).
- `experiments/p3_specialization/diagnostics/pairwise_action_kl.py` — **NEEDS UPDATE**: extend `softmax_packed` (lines 37-43) to include the signal head (10×2×2 = 40-class joint).
- `bucket_brigade/training/joint_trainer.py` — supports `centralized_critic` flag (line 213); action_dims default `[10, 2, 2]` (line 32).
- `bucket_brigade/training/networks.py` — `PolicyNetwork.action_heads` iterates `action_dims` (line 58); will auto-build 3 heads post-#236.
- `research_notebook/2026-05-16_issue239_post236_revalidation.md` — **new file** for side-by-side comparison.
- `~/.claude/projects/-Users-rwalters-GitHub-bucket-brigade/memory/project_ppo_failure_mode.md` — update only if verdicts flip.

## Test Plan

- [ ] Pre-flight smoke (4 quick cells, ~5 min): confirms action space is `MultiDiscrete([10, 2, 2])` end-to-end, action heads save with 3 keys, and the KL diagnostic returns finite values for the new joint.
- [ ] KL diagnostic fix: `pairwise_action_kl.py:softmax_packed` extended to include the signal head; unit-checked by running on a smoke cell and confirming the returned joint shape is `[T, 40]`.
- [ ] Obs-fix re-validation sweep (12 cells) completes; `analyze_220.py` emits `summary.{md,json}` with both scenarios populated.
- [ ] MAPPO re-validation sweep (18 cells) completes; `analyze_231.py` emits `summary.{md,json}` with headline verdict computed.
- [ ] Analyzer outputs use **post-#236 baselines** from #237/#238 (sanity-check the constants in the analyzers match what those PRs committed).
- [ ] Dated notebook entry compares pre-#236 (5.9% obs-fix; tier-3 MAPPO) vs post-#236 verdicts side-by-side.
- [ ] If any verdict flips: project memory updated with explicit pre/post numbers.

## References
- PR #236 (the substrate change — makes signaling first-class)
- PR #233 (issue #220 pre-#236 verdict: 5.9%)
- PR #232 (issue #231 pre-#236 verdict: tier-3, all 3 scenarios)
- PR #225 (centralized critic / MAPPO infrastructure)
- PR #216 (per-agent obs differentiation)
- Sibling issues #237 (random baseline re-derive), #238 (specialist baseline re-derive)
- Project memory: `project_ppo_failure_mode.md` (currently states MAPPO failed; may need correction)

## Acceptance criteria
- [ ] Pre-flight smoke tests pass (4 cells, 1 iter each)
- [ ] `pairwise_action_kl.py` updated to handle 3-head action space (10×2×2 = 40-class joint)
- [ ] Obs-fix re-validation sweep complete on post-#236 main (12 cells)
- [ ] MAPPO re-validation sweep complete on post-#236 main (18 cells)
- [ ] Both analyzers (`analyze_220.py`, `analyze_231.py`) re-run after #237/#238 commit updated baseline constants
- [ ] Verdicts applied per pre-registered ladder above
- [ ] Dated notebook entry comparing pre-#236 vs post-#236 outcomes side-by-side
- [ ] Project memory updated if verdicts flip

**Comment by rjwalters** [comment]
📝 **Curator** (2026-05-16): Comprehensive enrichment added covering checkpoint-compatibility analysis (obs_dim=46 invariant, action heads 2→3 NEW), required KL-diagnostic fix (`pairwise_action_kl.py` hardcodes 20-class joint; needs 40-class for [10,2,2]), tmux session naming, parallelism plan to coexist with sibling issues #237/#238, pre-flight smoke tests, gap_closed formula confirmation, run-dir layout, pre-registered verdict ladders for both sweeps, and risks. Dependencies section explicitly links #237/#238 as required for analyzer constants. Original issue text preserved at top.

**Comment by rjwalters** [comment]
Starting work on this issue per user request (override loom:curated). Will run pre-flight code fix to pairwise_action_kl.py first, then smoke test, then full sweep on COMPUTE_HOST_PRIMARY.
